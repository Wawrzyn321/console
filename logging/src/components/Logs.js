import React from 'react';
import PropTypes from 'prop-types';
import Header from './Header/Header';
import CompactHeader from './CompactHeader/CompactHeader';
import LogTable from './LogTable/LogTable';
// import 'core-js/es/array/flat-map'; todo

import {
  SORT_ASCENDING,
  DEFAULT_PERIOD,
  //LOG_REFRESH_INTERVAL,
} from './../constants';

const SAMPLE_LOGS = [
  {
    timestamp: '14:14:21.384196009Z',
    log: `[2019-06-11 11:58:00.047][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.1api.v2.Listener.use_original_dst' from file lds.proto. This configuration will be removed from 
  Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.`,
  },
  {
    timestamp: '14:14:01.384196009Z',
    log: `[2019-06-11 11:58:00.047][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.api.v2.Listener.use_original_dst' from file lds.proto. This configuration will be removed from 
  Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.`,
  },
  {
    timestamp: '14:14:31.384196009Z',
    log: `[2019-06-11 11:58:00.047][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.api.v2. details.`,
  },
  {
    timestamp: '14:14:41.384196009Z',
    log: `log`,
  },
  {
    timestamp: '14:14:51.384196009Z',
    log: `[2019-06-11 11:58:00.0147][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.api.v2.Listener.use_original_dst' from file lds.proto. This configuration will be removed from 
  Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.`,
  },
];

export default class Logs extends React.Component {
  static propTypes = {
    httpService: PropTypes.object.isRequired,
    queryTransformService: PropTypes.object.isRequired,
    isLambda: PropTypes.bool,
    readonlyLabels: PropTypes.arrayOf(PropTypes.string.isRequired),
    lambdaName: PropTypes.string,
  };

  static defaultProps = {
    isLambda: false,
    readonlyLabels: [],
    lambdaName: null,
  };

  state = {
    compact: this.props.isLambda,
    searchPhrase: '',
    labels: [],
    readonlyLabels: this.props.readonlyLabels,
    logsPeriod: DEFAULT_PERIOD,
    advancedSettings: {
      query: '',
      resultLimit: 100,
      showPreviousLogs: true,
      showHealthChecks: true,
    },
    sortDirection: SORT_ASCENDING,
    logs: SAMPLE_LOGS,
  };
  intervalId = null;

  componentDidMount = () => {
    const { labels, searchPhrase } = this.state;
    this.setState({
      advancedSettings: {
        ...this.state.advancedSettings,
        query: this.props.queryTransformService.toQuery(labels, searchPhrase),
      },
    });
    //this.intervalId = setInterval(this.fetchLogs, LOG_REFRESH_INTERVAL);
  };

  componentWillUnmount = () => {
    //clearInterval(this.intervalId);
  };

  fetchLogs = async () => {
    const {
      searchPhrase,
      labels,
      readonlyLabels,
      advancedSettings,
      sortDirection,
      logsPeriod,
    } = this.state;

    if (!labels.length && !readonlyLabels.length && !searchPhrase) {
      return;
    }

    const {
      resultLimit,
      showPreviousLogs,
      showHealthChecks,
    } = advancedSettings;

    try {
      const res = await this.props.httpService.fetchLogs({
        searchPhrase,
        labels: [...readonlyLabels, labels],
        resultLimit,
        logsPeriod,
        sortDirection,
        showPreviousLogs,
        showHealthChecks,
      });
      const logs = res.streams
        ? res.streams
            .flatMap(stream => stream.entries)
            .map(l => ({
              timestamp: l.ts,
              log: l.line,
            }))
        : [];
      this.setState({ logs });
    } catch (e) {
      console.warn(e);
    }
  };

  filterHealthChecks = entry => {
    const showHealthChecks = this.state.advancedSettings.showHealthChecks;
    return showHealthChecks || entry.log.indexOf('GET /healthz') < 0;
  };

  // intercept setState to update query, labels and searchPhrase
  updateState = partialState => {
    const { parseQuery, toQuery } = this.props.queryTransformService;
    const { labels, searchPhrase, query } = this.state;

    let additionalState = {};

    if ('searchPhrase' in partialState) {
      // searchPhrase changed, update query
      additionalState = {
        advancedSettings: {
          ...this.state.advancedSettings,
          query: toQuery(labels, partialState.searchPhrase),
        },
      };
    } else if ('labels' in partialState) {
      // labels changed, update query
      additionalState = {
        advancedSettings: {
          ...this.state.advancedSettings,
          query: toQuery(partialState.labels, searchPhrase),
        },
      };
    } else if (
      'advancedSettings' in partialState &&
      partialState.advancedSettings.query !== query
    ) {
      // query changed, update searchPhrase and labels
      additionalState = {
        ...this.state.advancedSettings,
        ...parseQuery(partialState.advancedSettings.query),
      };
    }

    this.setState({ ...partialState, ...additionalState });
  };

  render() {
    const {
      searchPhrase,
      labels,
      readonlyLabels,
      logsPeriod,
      sortDirection,
      advancedSettings,
      compact,
      logs,
    } = this.state;

    return (
      <>
        {compact ? (
          <CompactHeader
            updateFilteringState={this.updateState}
            searchPhrase={searchPhrase}
            logsPeriod={logsPeriod}
            sortDirection={sortDirection}
            advancedSettings={advancedSettings}
          />
        ) : (
          <Header
            updateFilteringState={this.updateState}
            searchPhrase={searchPhrase}
            labels={labels}
            readonlyLabels={readonlyLabels}
            logsPeriod={logsPeriod}
            sortDirection={sortDirection}
            advancedSettings={advancedSettings}
          />
        )}
        <LogTable
          entityName={this.props.lambdaName}
          entries={logs.filter(this.filterHealthChecks)}
        />
      </>
    );
  }
}
