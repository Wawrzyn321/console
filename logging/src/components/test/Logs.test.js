import React from 'react';
import renderer from 'react-test-renderer';
import Logs from './../Logs';

describe('Logs', () => {
  it('Renders with minimal props', () => {
    console.error = jest.fn();

    afterEach(() => {
      console.error.mockReset();
    });

    afterAll(() => {
      expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
    });

    const component = renderer.create(
      <Logs httpService={{}} queryTransformService={{ toQuery: () => {} }} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders with lambda view props', () => {
    console.error = jest.fn();

    afterEach(() => {
      console.error.mockReset();
    });

    afterAll(() => {
      expect(console.error.mock.calls[0][0]).toMatchSnapshot(); // unique "key" prop warning
    });

    const component = renderer.create(
      <Logs
        httpService={{}}
        queryTransformService={{ toQuery: () => {} }}
        isLambda={true}
        readonlyLabels={[]}
        lambdaName="test"
      />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
