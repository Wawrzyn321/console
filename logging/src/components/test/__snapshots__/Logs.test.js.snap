// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logs Renders with lambda view props 1`] = `
Array [
  <div
    className="fd-panel fd-has-padding-regular fd-has-padding-bottom-tiny"
  >
    <section
      className="compact-header"
    >
      <h1
        className="fd-has-type-3"
      >
        Logs
      </h1>
      <div
        className="fd-has-display-flex"
      >
        <section
          className="fd-has-margin-right-small"
        >
          <span
            className="caption-muted search-input__caption"
          />
          <input
            className="search-input--compact"
            id="search-input"
            min="0"
            onChange={[Function]}
            placeholder="Search"
            type="text"
            value=""
          />
        </section>
        <span
          className="link-button fd-has-type-minus-1"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <button
                  className="fd-button--light sap-icon--action-settings fd-has-margin-right-tiny"
                  onClick={[Function]}
                  size="xs"
                />
              </div>
            </div>
          </div>
        </span>
        <span
          className="link-button fd-has-type-minus-1 select-dropdown"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <button
                  className="fd-button--light sap-icon--past fd-has-margin-right-tiny"
                  onClick={[Function]}
                  size="xs"
                />
              </div>
            </div>
          </div>
        </span>
        <span
          className="link-button fd-has-type-minus-1 select-dropdown"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <button
                  className="fd-button--light sap-icon--sort fd-has-margin-right-tiny"
                  onClick={[Function]}
                  size="xs"
                />
              </div>
            </div>
          </div>
        </span>
      </div>
    </section>
  </div>,
  <table
    className="fd-table fd-has-margin-regular"
  >
    <thead>
      <tr>
        <th
          className="log-table__pre-header fd-has-type-0"
          colSpan="2"
        >
          Logs for test
        </th>
      </tr>
      <tr>
        <th
          className="caption-muted"
        >
          Timestamp
        </th>
        <th
          className="caption-muted"
        >
          Log
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:21.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          [2019-06-11 11:58:00.047][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.1api.v2.Listener.use_original_dst' from file lds.proto. This configuration will be removed from 
  Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.
        </td>
      </tr>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:01.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          [2019-06-11 11:58:00.047][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.api.v2.Listener.use_original_dst' from file lds.proto. This configuration will be removed from 
  Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.
        </td>
      </tr>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:31.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          [2019-06-11 11:58:00.047][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.api.v2. details.
        </td>
      </tr>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:41.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          log
        </td>
      </tr>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:51.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          [2019-06-11 11:58:00.0147][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.api.v2.Listener.use_original_dst' from file lds.proto. This configuration will be removed from 
  Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.
        </td>
      </tr>
    </tbody>
  </table>,
]
`;

exports[`Logs Renders with minimal props 1`] = `
Array [
  <div
    className="fd-panel fd-has-padding-regular fd-has-padding-bottom-none"
  >
    <h1
      className="fd-has-type-3 fd-has-padding-bottom-tiny"
    >
      Logs
    </h1>
    <section
      className="header__settings-group"
    >
      <section
        className="fd-display-l-inline-block"
      >
        <span
          className="caption-muted"
        >
          Labels
        </span>
        <div
          className="fd-combobox-input"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <div
                  className="fd-combobox-control"
                  onClick={[Function]}
                >
                  <div
                    className="fd-input-group fd-input-group--after"
                  >
                    <input
                      className="fd-input"
                      placeholder="Select Label"
                      type="text"
                    />
                    <span
                      className="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button"
                    >
                      <button
                        className=" fd-button--light sap-icon--navigation-down-arrow"
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
      <section
        className="fd-has-margin-right-small"
      >
        <span
          className="caption-muted search-input__caption"
        >
          Search
          <div
            className="small-inline-help-wrapper"
          >
            <span
              className="fd-inline-help"
            >
              <span
                className="fd-inline-help__content fd-inline-help__content--right"
              >
                Search for logs by text (optional)
              </span>
            </span>
          </div>
        </span>
        <input
          className=""
          id="search-input"
          min="0"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </section>
      <span
        className="link-button fd-has-type-minus-1 header__settings-group__toggle"
        data-test-id="advanced-settings-toggle"
        onClick={[Function]}
      >
        Show Advanced Settings
      </span>
    </section>
    <div>
      <section
        className="labels-display"
      >
        <ul />
      </section>
      <div
        className="header__options-wrapper"
      >
        <span
          className="link-button fd-has-type-minus-1 select-dropdown"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <button
                  className="fd-button--light sap-icon--past fd-has-margin-right-tiny"
                  onClick={[Function]}
                  size="xs"
                >
                  last 15 minutes
                </button>
              </div>
            </div>
          </div>
        </span>
        <span
          className="link-button fd-has-type-minus-1 select-dropdown"
        >
          <div
            className="fd-popover"
          >
            <div
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                className="fd-popover__control"
              >
                <button
                  className="fd-button--light sap-icon--sort fd-has-margin-right-tiny"
                  onClick={[Function]}
                  size="xs"
                >
                  ascending
                </button>
              </div>
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>,
  <table
    className="fd-table fd-has-margin-regular"
  >
    <thead>
      <tr>
        <th
          className="log-table__pre-header fd-has-type-0"
          colSpan="2"
        >
          Log stream
        </th>
      </tr>
      <tr>
        <th
          className="caption-muted"
        >
          Timestamp
        </th>
        <th
          className="caption-muted"
        >
          Log
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:21.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          [2019-06-11 11:58:00.047][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.1api.v2.Listener.use_original_dst' from file lds.proto. This configuration will be removed from 
  Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.
        </td>
      </tr>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:01.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          [2019-06-11 11:58:00.047][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.api.v2.Listener.use_original_dst' from file lds.proto. This configuration will be removed from 
  Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.
        </td>
      </tr>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:31.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          [2019-06-11 11:58:00.047][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.api.v2. details.
        </td>
      </tr>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:41.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          log
        </td>
      </tr>
      <tr>
        <td
          className="caption-muted"
        >
          14:14:51.384196009Z
        </td>
        <td
          className="caption-muted"
        >
          [2019-06-11 11:58:00.0147][15][warning][misc] [external/envoy/source/common/protobuf/utility.cc:174] 
  Using deprecated option 'envoy.api.v2.Listener.use_original_dst' from file lds.proto. This configuration will be removed from 
  Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.
        </td>
      </tr>
    </tbody>
  </table>,
]
`;
